### PR TITLE
Switch to Play 3 and go all in on Pekko

### PR DESCRIPTION
**Pull request by mkurz**
Hi!

Play maintainer here :wink: I see you switched most of the repo to Pekko already:
- #26563
- #26602

Not sure if you have seen it, we finally (thanks to your support!) released Play 3, which starts to use Pekko under the hood, more can be found here:
- https://twitter.com/playframework/status/1721576472047911007
- https://github.com/playframework/playframework/releases/tag/3.0.0

All the new features, release highlights and the migration guide from Play 2.9 apply to 3.0 as well:
- Highlights 2.9: https://www.playframework.com/documentation/latest/Highlights29
- Highlights 3.0: https://www.playframework.com/documentation/latest/Highlights30 ("just" the Pekko switch + groupId change)
- Migration guide for 2.9: https://www.playframework.com/documentation/latest/Migration29
- Migration guide for 3.0: https://www.playframework.com/documentation/latest/Migration30 (again "only" pekko + groupId)

I took a look at your repo and started the migration; it's probably not all that needs to be done, but to get started. I will try my best to help to get things working for you. I will see what fails and push more commits to this pull request.
E.g. it might be possible that webjars cause problems, since Play 3 uses sbt-web 1.5 now which was refactored a bit. We will see. Also I think some dependencies need to be upgraded, like https://github.com/bizzabo/play-json-extensions (?). If so, I am happy to help there also.

If you have any further questions I am available.

BTW: Thanks for your ongoing financial support for Play, without it we would *definitely* not have been able to push out those releases, thank you very much!

### CODE REVIEW

File: common/app/http/Filters.scala [hunk 1]
```diff
@@ -144,7 +144,6 @@ object Filters {
 
 }
 
-//Note: still using akka (instead of pekko) materializer from Play as both filters extend Play's Filter, so the types need to match
```
Comment (mkurz) [inline common/app/http/Filters.scala:147]: Is it ok to just remove that comment now? I didn't dig deeper if something else needs to be done.

File: project/plugins.sbt [hunk 2]
```diff
@@ -7,38 +7,18 @@ libraryDependencies ++= Seq(
...
-addSbtPlugin("com.timushev.sbt" % "sbt-updates" % "0.3.3")
+addSbtPlugin("com.timushev.sbt" % "sbt-updates" % "0.6.3")
 
-addSbtPlugin("com.typesafe.sbt" % "sbt-git" % "0.9.3")
+addSbtPlugin("com.github.sbt" % "sbt-git" % "2.0.1")
 
 addSbtPlugin("org.scalameta" % "sbt-scalafmt" % "2.4.0")
 
-addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.10.0-RC1")
+addDependencyTreePlugin
```
Comment (mkurz) [inline project/plugins.sbt:22]: sbt-dependency-grap is now build-in in sbt, see https://www.scala-sbt.org/1.x/docs/sbt-1.4-Release-Notes.html#sbt-dependency-graph+is+in-sourced

File: project/plugins.sbt [hunk 3]
```diff
@@ -7,38 +7,18 @@ libraryDependencies ++= Seq(
   "org.joda" % "joda-convert" % "1.7",
 )
 
-resolvers ++= Resolver.sonatypeOssRepos("releases") ++ Seq(
-  Classpaths.typesafeReleases,
-  Resolver.typesafeRepo("releases"),
-)
```
Comment (mkurz) [inline project/plugins.sbt:13]: These typesafe resolver _should_ not be needed anymore :crossed_fingers:

File: project/plugins.sbt [hunk 4]
```diff
@@ -7,38 +7,18 @@ libraryDependencies ++= Seq(
...
+resolvers ++= Resolver.sonatypeOssRepos("releases")
 
-addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.8.19")
+addSbtPlugin("org.playframework" % "sbt-plugin" % "3.0.0")
 
-addSbtPlugin("com.typesafe.sbt" % "sbt-native-packager" % "1.3.1")
+addSbtPlugin("com.github.sbt" % "sbt-native-packager" % "1.9.16")
```
Comment (mkurz) [inline project/plugins.sbt:14]: Actually not sure if sbt-native-packager needs to be pulled in explicitly because Play should provide it anyway.

**Review by AshCorr — COMMENTED** [review]
Hey @mkurz !

Wow! We were not expecting someone else to do this migration for us, thank you so much! We really _**really**_ appreciate all the work you and the rest of the Play framework team put into maintaining such a great library!

As the group ID for the Play framework is changing to `org.playframework` now will we need to make sure that all of our dependencies that use Play also use the new group ID?

I'm thinking these 3 in particular which we maintain https://github.com/guardian/frontend/pull/26677/files#diff-ad2642dc77b3679e4ad2c7d3b2f59a2dcf48c1bf5b93a558a4fa44828315a34cL54-L56

**Comment by mkurz** [comment]
> Wow! We were not expecting someone else to do this migration for us, thank you so much!

We probably need to do this together :stuck_out_tongue_closed_eyes:

> We really _**really**_ appreciate all the work you and the rest of the Play framework team put into maintaining such a great library!

:heart: 

> As the group ID for the Play framework is changing to `org.playframework` now will we need to make sure that all of our dependencies that use Play also use the new group ID?

Yes, all dependencies need to switch their own dependencies to `org.playframework` now. There could be exceptions where it does not matter, but it's definitely much safer if everything switches.

I was looking at the libs you mentioned: It shouldn't be to hard to migrate them. I am somehow motivated to pick this up. However there are some things I would change if you don't mind:

1. would it be ok if you first delete all those unrelevant, old branches which are "closed" (red) and "merged" (purble) from both repos, just to do a bit of cleanup: 
    - https://github.com/guardian/play-googleauth/branches/all
    - https://github.com/guardian/play-secret-rotation/branches/all
1. I see that you pull in v2.2.7 of the play-googleauth lib, but there is no git-tag for that:
     - https://github.com/guardian/frontend/blob/3366a31f827fa8b14ff7f0a38e58740a8ab76c33/project/Dependencies.scala#L54
    - https://github.com/guardian/play-googleauth/tags
1. I see that in those projects you use an approach of creating a folder for each major Play release, like [here](https://github.com/guardian/play-googleauth/tree/main/play-v28), where you than just link the the play-v27 folder. Or [here](https://github.com/guardian/play-secret-rotation/tree/main/play/) you link even both, 2.7 and 2.8, to 2.6. I would suggest to not do that, because e.g. now in play-googleauth, where you [`import akka.*`](https://github.com/guardian/play-googleauth/blob/5315e8b0498ff9ac7b531769c6e2ae4923ed66c0/play-v27/src/main/scala/com/gu/googleauth/filters.scala#L3), the setup would get even more complex, with maybe duplicated files, just to switch that import to `org.apache.pekko`. IMHO it would be better if each major play release would be maintained in its own branch. That could be done by naming the branches e.g. 1.x which contains the Play 2.7 code base, then for Play 2.8.x the branch would 2.x, and so on (this is just an example). Or name the branches after the play releases (`play28` for Play 2.8, etc). Again just an example. However that would avoid of doing this folder-per-release thing with symlinking and instead back-/forwardporting bug fixes and features between branches which is IMHO much nicer and the setup simpler. What I would do now is:
    1. Create a branch on current HEAD of both repos, named `play2` or similiar, to mark this is the branch that contains the (now "old") Play 2.x code base of those libs. Then I would start to clean up the main branch, by removing all that symlink stuff to have a nice setup. The next play-googleauth release would then be 3.0.0, from the master branch. For a future Play 3.1 release  you could then branch off a 3.0.x branch from main and migrate to Play 3.1 in the main and so on. You could than always backport changes from main to 3.0.x if desired and cut release from that branch for Play 3.0 etc.

What do you think about that?

**Comment by johnduffell** [comment]
This is really helpful, thanks @mkurz as I was just starting this yesterday for https://github.com/guardian/members-data-api
I have just picked up a couple of tips from your PR and I will continue today.
The good thing is upgrading to play 2.9 or 3.0 will unlock the scala 3 upgrade path, and all the readability and learning curve benefits that scala 3 brings!

**Comment by mkurz** [comment]
BTW, I guess you still run on Java 11? I recommend that you switch to Java 17, or even better Java 21 already (both are LTS). We plan to drop Java 11 in Play 3.1.0 (ETA somehow next summer or autumn, so like roughly 1 year from now). However very likely just by switching to Java 17/21 you might see a performance boost (?), at least that is what we observed in benchmarks and one Play Framework sponsor reported that to me.

**Comment by AshCorr** [comment]
> would it be ok if you first delete all those unrelevant, old branches which are "closed" (red) and "merged" (purble) from both repos, just to do a bit of cleanup

Can do, we tend not to use forks so we do end up with a fair few stale branches on our repos.

> I see that you pull in v2.2.7 of the play-googleauth lib, but there is no git-tag for that
> I see that in those projects you use an approach of creating a folder for each major Play release

Yea we could probably do with rethinking how we manage our various versions of these dependencies, our current approach doesn't seem that sustainable. Using tags/branches to keep track of older versions of the library feels like it could make upgrades a fair bit easier.

> BTW, I guess you still run on Java 11?

This repo does, our recommendation is to use Java 17 but we haven't gotten around to upgrading everything yet.

**Comment by ioannakok** [comment]
Hi @mkurz, thanks so much for your contribution! As others have already said, we really appreciate it. I have re-raised your PR here:
* https://github.com/guardian/frontend/pull/26755
because in order for the tests to run someone from the Guardian organisation has to raise it. I hope that's okay! I am going to add a comment there with an update.

**Pull request closed without merging**